L22-change-grpc++-dir-name.md

```diff
@@ -0,0 +1,43 @@
+gRPC C++ Public Header Directory Change
+----
+* Author(s): muxi
+* Approver: nicolasnoble
+* Status: draft
+* Implemented in: C++
+* Last updated: 01/25/2018
+* Discussion at: https://groups.google.com/d/msg/grpc-io/SDyc0hSWWG8/pX2n9BSNAQAJ
+
+## Abstract
+This proposal is to change the name of directory `include/grpc++` to `include/grpcxx` for compatibility issues. 
+
+## Background
+gRPC C++ headers have been using the directory `include/grpc++` for a long time. Currently, users use `#include <grpc++/...>` to include gRPC headers in their source code. Some gRPC internal code and public headers also use this style to include gRPC public headers.
+
+However, since the name `grpc++` has special character `+`, it causes compatibility issue when migrating to certain other platforms.
+
+One of such example is Xcode. Some gRPC users, such as Firestore, need to build gRPC C++ library on iOS as Apple framework. The problem emerges on this platform due to 3 restrictions:
+- gRPC public headers use `#include <grpc++/...>` to include gRPC C++ headers;
+- An iOS app must include headers of a Framework with the format `#include <framework_name/path_to_header/header.h>`;
+- Apple framework's name must be C99 extended identifier compatible; `grpc++` is not compatible.
+
+The 3 restrictions make it impossible to make gRPC C++ library work as Apple framework. We believe this issue can happen somewhere else too in the future (e.g. [grpc#14089](https://github.com/grpc/grpc/pull/14089) is another example where this naming convention creates problem).
+
+## Proposal
+
+The proposal is to migrate `include/grpc++` directory to `include/grpcxx` in a backwards compatible manner. The objective is that both styles of inclusion `#include <grpc++/...>` and `#include <grpcxx/...>` can be used for all past and future gRPC C++ users.
+
+### Changes to gRPC C++ API
+
+- Make `include/grpcxx` directory the new location for all gRPC C++ public headers; move all current headers from `include/grpc++` to `include/grpcxx`; update all corresponding inclusions inside gRPC code base;
+- Make wrapper headers in `include/grpc++` for all C++ public headers, preserving the directory structure. This makes current gRPC users build with old directory name.
```

Comment:
Maybe make sure there is some sort of deprecation warning if using these headers ? Not sure if that's a good idea or not, but, relevant, maybe.

Reply:
Yep. I think it is good adding deprecation warning as comment at the top of the wrapper headers.

Reply:
Can I suggest that that should be considered breaking? A lot of people mention using -Werror . Can it be a message that's not a warning? I don't know how to implement that.

Reply:
`#pragma message`?

Reply:
I am not very sure how portable that is though. We can give a try.